clang/lib/AST/Decl.cpp

```diff
@@ -5031,7 +5031,12 @@ void RecordDecl::completeDefinition() {
 /// This which can be turned on with an attribute, pragma, or the
 /// -mms-bitfields command-line option.
 bool RecordDecl::isMsStruct(const ASTContext &C) const {
-  return hasAttr<MSStructAttr>() || C.getLangOpts().MSBitfields == 1;
+  if (hasAttr<MSStructAttr>())
+    return true;
+  if (hasAttr<GCCStructAttr>())
+    return false;
+  return C.getLangOpts().MSBitfields.value_or(
```

Comment:
Is there a reason we don't just set the lang opt correctly for the target?  I'm pretty sure language option defaults depending on target settings is well-established.

Reply:
But there is no `TargetInfo` object when I set it in `CompilerInvocation::ParseLangArgs`.

Reply:
You have a target triple, and you have the base CXXABI. What information are you looking to get from the `TargetInfo`?

Reply:
I still don't quite understand what you want.

Yes, I can compute default value while parsing arguments, but I still will need `TargetInfo::defaultsToMsStruct`, since I use it in SemaDeclCxx.

Reply:
I always prefer to have less complicated logic in the options types, because I've seen it get out of hand.  I don't feel too strongly about this, though, and I do see how you have uses of the components feeding into it.